alibi/explainers/similarity/base.py

```diff
@@ -0,0 +1,174 @@
+            for X, Y in tqdm(zip(self.X_train, self.Y_train)):
+                grad_X_train = self._compute_grad(X[None], Y[None])
+                grads.append(grad_X_train[None])
+            self.grad_X_train = np.concatenate(grads, axis=0)
+        return self
+
+    def _verify_fit(self) -> None:
+        """Verify that the explainer has been fitted.
+
+        Raises
+        ------
+        ValueError
+            If the explainer has not been fitted.
+        """
+
+        if not hasattr(self, 'X_train') or not hasattr(self, 'Y_train'):
+            raise ValueError('Training data not set. Call `fit` and pass training data first.')
+
+    def _match_shape_to_data(self,
+                             data: 'Union[np.ndarray, tensorflow.Tensor, torch.Tensor]',
+                             target_type: Literal['X', 'Y']) -> 'Union[np.ndarray, tensorflow.Tensor, torch.Tensor]':
+        """Verify the shape of `data` against the shape of the training data.
+
+        Used to ensure input is correct shape for gradient methods implemented in the backends. `data` will be the
+        features or label of the instance being explained. If the `data` is not a batch, reshape to be a single batch
+        element. i.e. if training data shape is `(3, 28, 28)` and data shape is `(3, 28, 28)` we reshape data to
+        `(1, 3, 28, 28)`.
+
+        Parameters
+        ----------
+        data
+            Data to be matched shape-wise against the training data.
+        target_type
+            Type of data: ``'X'`` | ``'Y'``. Used to determine if data should take the shape of predictor input or
+            predictor output. ``'X'`` will utilize the `X_dims` attribute which stores the shape of the training data.
+            ``'Y'`` will match the shape of `Y_dims` which is the shape of the target data.
+
+        Raises
+        ------
+        ValueError
+            If the shape of `data` does not match the shape of the training data, or fit has not been called prior to
+            calling this method.
+        """
+        target_shape = getattr(self, f'{target_type}_dims')
+        if data.shape == target_shape:
+            data = data[None]
+        if data.shape[1:] != target_shape:
+            raise ValueError((f'Input `{target_type}` has shape {data.shape[1:]}'
+                              f' but training data has shape {target_shape}'))
+        return data
+
+    def _compute_adhoc_similarity(self, grad_X: np.ndarray) -> np.ndarray:
+        """
+        Computes the similarity between the gradients of the test instances and all the training instances. The method
+        performs the computation of the gradients of the training instance on the fly without storing them in memory.
+
+        parameters
+        ----------
+        grad_X
+            Gradients of the test instances.
+        """
+        scores = np.zeros(self.X_train.shape[0])
+        for i, (X, Y) in tqdm(enumerate(zip(self.X_train, self.Y_train))):
```

Comment:
I am thinking that maybe we can use tqdm only when a verbose argument is set to `True`. Just an idea ...